### ISSUE TITLE
NMODL and MOD2C pass different weight indices during initialisation

### ISSUE THREAD

**Issue by olupton** [BlueBrain/nmodl#680]
In the `nrn_init_...` method generated by NMODL for, for example, [Gfluct3.mod](https://github.com/neuronsimulator/testcorenrn/blob/7b781b3a7ac1e0b785f897e4db5c3bf84c5127d2/mod/Gfluct3.mod) there is a call
```c++
net_send_buffering(ml->_net_send_buffer, 0, inst->tqitem[2*pnodecount+id], -1, inst->point_process[1*pnodecount+id], nt->_t+0.0, 1.0);
```
where the 4th argument is `-1`, but in the MOD2C equivalent `0` is passed. This does not appear to have any effect on whether the tests pass, but it would be good to understand if this difference is a bug or not.

**Comment by alkino** [comment]
I think here we should get attention form @nrnhines

**Comment by pramodk** [comment]
@alkino : this is NMODL bug. We need to compare with MOD2C implementation and find out why/how NMODL doing it differently.

**Comment by pramodk** [comment]
Looking into MOD2C for [_net_send_buffering](https://github.com/BlueBrain/mod2c/blob/9b474e903028cf5512f6998c428fa03c4f3c326f/src/mod2c_core/nocpout.c#L3033) function:

```cpp
static void _net_send_buffering(NetSendBuffer_t* _nsb, int _sendtype, int _i_vdata, int _weight_index, int _ipnt, double _t, double _flag)
```
the 4th parameter is `weight_index`. If I look into how the `_net_send_buffering()` is printed in mod2c at [this](https://github.com/BlueBrain/mod2c/blob/9b474e903028cf5512f6998c428fa03c4f3c326f/src/mod2c_core/nocpout.c#L3135) :

```cpp
_net_send_buffering(_ml->_net_send_buffer, 0, _tqitem, 0, _ppvar[1*_STRIDE], t +  0.0 , %s );\n"
```

and [this](https://github.com/BlueBrain/mod2c/blob/9b474e903028cf5512f6998c428fa03c4f3c326f/src/mod2c_core/parsact.c#L1064) point

```cpp
void acc_net_add(Item * qname, Item * qpar1, Item * qexpr, Item * qpar2,
        const char * xarg1,
            const char * xarg2) {
        Item * q;
        if (artificial_cell) {
            return;
        }
        sprintf(buf, "\
                n _net_send_buffering(_ml -> _net_send_buffer, % s\ ", xarg1);

.....

			if (blocktype == NETRECEIVE) {
			    acc_net_add(qname, qpar1, qexpr, qpar2, "0, _tqitem, _weight_index, _ppvar[1*_STRIDE],", "");
			    Insertstr(qpar1 -> next, "_tqitem, _weight_index, _pnt,");
			} else if (blocktype == INITIAL1) {
			    net_send_buffer_in_initial = 1;
			    acc_net_add(qname, qpar1, qexpr, qpar2, "0, _tqitem, 0,  _ppvar[1*_STRIDE],", "");
			    Insertstr(qpar1 -> next, "_tqitem, -1, (Point_process*) _nt->_vdata[_ppvar[1*_STRIDE]],");
			} else {
			    diag("net_send allowed only in INITIAL and NET_RECEIVE blocks", (char * ) 0);
			}
			} else if (strcmp(SYM(qname) -> name, "net_event") == 0) {
			    net_event_seen_ = 1;
			    if (blocktype == NETRECEIVE) {
			        acc_net_add(qname, qpar1, qexpr, qpar2, "1, -1, -1,  _ppvar[1*_STRIDE],", ", 0.");
			        Insertstr(qpar1 -> next, "_pnt,");
			    } else {
			        diag("net_event", " only allowed in NET_RECEIVE block");
			    }
			} else if (strcmp(SYM(qname) -> name, "net_move") == 0) {
			    if (artificial_cell) {
			        replacstr(qname, "artcell_net_move");
			    }
			    if (blocktype == NETRECEIVE) {
			        acc_net_add(qname, qpar1, qexpr, qpar2, "2, _tqitem, -1, _ppvar[1*_STRIDE],", ", 0.");
			        Insertstr(qpar1 -> next, "_tqitem, _pnt,");
			    } else {
			        diag("net_move", " only allowed in NET_RECEIVE block");
			    }
			}
```
The summary from above is:

*  If we have `net_send` event then
   * `weight_index` is used for `NET_RECEIVE` block
   * `0` is used for `INITIAL` block
   * all others are error because `net_send allowed only in INITIAL and NET_RECEIVE blocks`
*  If we have `net_event` event then
   * `-1` is used for `NET_RECEIVE` block
   * all others are error because `net_event  only allowed in NET_RECEIVE block`
* If we have `net_move` event then
  * `-1` is used for `NET_RECEIVE` block
  * all others are error because `net_move  only allowed in NET_RECEIVE block`

**Comment by pramodk** [comment]
A slight correction to the 1st point in https://github.com/BlueBrain/nmodl/issues/680#issuecomment-885564763:
```
If we have net_send event then
 - weight_index is used for NET_RECEIVE block
 - weight_index is used for INITIAL block within NET_RECEIVE block see (bug fix #1017) -- NEW**
 - 0 is used for top-level INITIAL block
 - all others are error because net_send allowed only in INITIAL and NET_RECEIVE blocks
```
